### DISCUSSION TITLE
Calling UseEndpoints multiple times from multiple IStartupFilters

### DISCUSSION THREAD

**Discussion by kimbell** [dotnet/aspnetcore#27503] [Q&A]
We are building up a collection of .NET Core services. We also have internal libraries that may add middleware. As this collection grows, one of the things we want to maintain is a consistent setup on how middleware is configured; this might change over time. 

`IStartupFilter `is one way we automatically add middleware. I've been looking around on the Internet, and explored the code from Steeltoe. In their codebase they are using `IStartupFilter `to call `UseEndpoints()`. Depending on what has been added to the `IServiceCollection`, the `UseEndpoints()` will be called multiple times. 

https://github.com/SteeltoeOSS/Steeltoe/blob/b3928ed4ec527be9fc07e917a615bbebad22c254/src/Management/src/EndpointCore/Loggers/LoggersStartupFilter.cs

https://github.com/SteeltoeOSS/Steeltoe/blob/b3928ed4ec527be9fc07e917a615bbebad22c254/src/Management/src/EndpointCore/Mappings/MappingsStartupFilter.cs

I haven't seen this approach anywhere else, so is this a supported option? 

Are there other options I can explore, or do I need to create my own framework specific for our needs?

**Comment by Tratcher** [comment]
IStartupFilter isn't a great middleware composition model because it gives very little control over ordering, and in this case duplication. Calling UseEndpoints multiple times isn't ideal because it will add multiple EndpointMiddleware to the pipeline.
https://github.com/dotnet/aspnetcore/blob/8c8cb4dce235b099d1e419b6a039217eb56a77e6/src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs#L105

I'm not aware of a good built-in solution for this scenario. Do you control the primary Startup class and Configure method? If so you should be able to provide your own interface like `IEndpointSetup.ConfigureEndpoints(IEndpointRouteBuilder builder)`, inject `IEnumerable<IEndpointSetup>` into Startup.Configure, and then call them inside the UseEndpoints configuration lambda. You could do something similar for inserting other middleware at designated points in your pipeline.

**Reply by kimbell** [reply]
I control the startup class and have similar ideas on how to solve this. Since I saw this pattern in a library, I just wanted to check if there was something I had missed.
